Port to Python3

harshkasat opened the pull request:
Initial changes were made to Article.py and Article_Data. These are simple changes; as I delve deeper into the repository, I will make necessary modifications.

Initial testing has been conducted confirming the functionality of Infoslicer with the embedded BeautifulSoup4 library and py3
- [x] Debian: Successful execution and data extraction verified.
- [ ] Fedora: Testing on Fedora is pending and will be completed in a subsequent update.
![done on sugar live (debian) ](https://github.com/user-attachments/assets/84bf0ff6-9459-4ab8-81ef-66982e21befa)

harshkasat committed:
inital changes done on Article.py and Article_Data these are simples change as i go deeper in repo i will make required change

MostlyKIGuess commented on infoslicer/processing/Article.py:
```diff
@@ -1,14 +1,15 @@
 import logging
 
-logger = logging.getLogger('infoslicer')
+# logger = logging.getLogger('infoslicer')
```
we can just remove this

harshkasat replied:
I think we need this; it will save logs, so if any error occurs, we can see it in the log file. I'm not using it yet, but I'll implement it in the future. Right now, I'm delving deeper into the repository.

MostlyKIGuess commented on infoslicer/processing/Article.py:
```diff
@@ -199,94 +171,93 @@ def checkIntegrity(self):
-        Likewise, if only paragraphs are being inserted, then the first sentence array will contain a dummy object.
-    
-        The section objects array and the second paragraph and sentence arrays, can all be empty.
+        Objects can have the following structure:
```
maybe we don't need this new help section

MostlyKIGuess commented on infoslicer/processing/Article.py:
```diff
@@ -199,94 +171,93 @@ def checkIntegrity(self):
         elif secend.get_offset() - lociter.get_offset() < 4:
             extra = 3    
-        
+
```
in this -20 , +20 lines lots of unwanted new lines and comments

harshkasat replied:
i think we need to write a pep8 or pylint test to remove this unwanted space

MostlyKIGuess replied:
Yeah could also do that, you can always run flake8 checks to avoid this

harshkasat replied:
@quozl  what are your thoughts on this? I think this can help us in the long run. Adding a GitHub Action after a PR is raised to check if they pass pytest and lint will be beneficial.

quozl replied:
Certainly not.  My threshold for using GitHub actions in a repository is where we are getting more than ten pull requests _per day_, and have more than 20 reviewers and committers.  This activity is very unlikely to ever reach that kind of milestone.

Just run your autopep8 or whatever and commit the changes in separate commits, so that we don't have to review automated changes.  

Remember that review of commits for automated changes isn't the same as for human changes; a reviewer would reset their repository to the hash before the automated change, issue the same command, and then compare against the other developer's change.  The commit message should explain the way the change was done.

For the threat of supply chain attack, for critical software components, we would visually check the change.  This activity _is certainly not_ a critical component.

MostlyKIGuess commented on infoslicer/processing/Article.py:
```diff
@@ -296,56 +267,59 @@ def insert(self, objects, lociter):
         """
         This method is the same as the above insert method, except that sentence objects are not included.
-    
+        
```
we shouldn't have trailing spaces \t can be avoided

MostlyKIGuess commented on infoslicer/processing/Article_Data.py:
This is nicely done, LGTM

MostlyKIGuess reviewed (commented):
Overall LGTM, maybe there is  extra changes that we don't need

MostlyKIGuess commented:
Not tested.. Waiting for more files

quozl commented:
Mark this as not draft when the activity runs properly.

You must also run the activity before the port under a Python 2 environment so that you  know what you are aiming for

harshkasat commented:
@quozl  https://github.com/sugarlabs/infoslicer/blob/master/infoslicer/processing/BeautifulSoup.py This is the source code of BeautifulSoup 3. Currently, BeautifulSoup 4 (bs4) is available and compatible with py3. What I proposed is that instead of porting to py3, we would use the bs4 module directly? It's actively maintained and provides better support. this can also resolve this issue: https://github.com/sugarlabs/infoslicer/issues/29

quozl commented:
I don't think you should need this guidance.  For any dependency, you may either remove it from inside the software if you explain briefly how to resolve the dependency in the README.md, or you may vendor it into the software if you give good reasons in your commit message.  You should consider how Sugar activity bundle delivery, Fedora packaging, Debian packaging, or Flatpak will be affected.  As this activity has taken so long to port, I think it has dropped off the activity library, Fedora and Debian distributions, but that doesn't make it unhelpful to understand what your decisions will do.

harshkasat commented:
@quoz and @MostlyKIGuess l  I need help!  how can I resolve it, so it occurs after I complete port py3 ` path infoslicer/processing`
![image](https://github.com/user-attachments/assets/6efd037c-7aec-4a91-a2ef-a39bd8de9d9f)

MostlyKIGuess commented:
> @quoz and @MostlyKIGuess l  I need help!  how can I resolve it, so it occurs after I complete port py3 ` path infoslicer/processing`
> ![image](https://github.com/user-attachments/assets/6efd037c-7aec-4a91-a2ef-a39bd8de9d9f)
> 


Can you also give more information and send error in text format on matrix, I would like to know what environment you are on and how exactly are you testing it, along with OS.

Seems like you are using gtktoolkit3 so you are checking the port I suppose and not pre port?

harshkasat commented:
@MostlyKIGuess It's weird! After taking a coffee break, it starts working perfectly.
![image](https://github.com/user-attachments/assets/6fdaba31-8025-466e-bbf3-8134eef5ef27)

harshkasat committed:
feat: implement core article handling

- Add Article class with GTK text buffer management
- Implement article data structures
- Add DITA parsing and generation

harshkasat committed:
remove whitespace, pylint error, error handlin

The changes fix:
- Fix the trailing whitespace
- Pylint Error handling
- add try and except for error handling
- remove built-in variable name

harshkasat commented:
@quozl I have question, so bs4 need dependencies `typing_extension`  so I come up with a setup.py i need know if its correct or 

- previous `setup.py`
```python
from sugar3.activity import bundlebuilder
if __name__ == "__main__":
    bundlebuilder.start()
``` 

- after `setup.py`

```python
import sys
import subprocess
from setuptools import setup, find_packages

def install_missing_dependencies():
    required = ['typing_extensions']
    for package in required:
        try:
            __import__(package)
        except ImportError:
            print(f"Installing missing dependency: {package}")
            subprocess.check_call([sys.executable, "-m", "pip", "install", package])

def main():
    install_missing_dependencies()
``` 

also I not 100% sure that we need  `bundlebuilder.start()` becoz `setup.py` never called in activity.py so may be we just remove it ?

quozl commented:
No, we do not automatically resolve dependencies in our setup.py files in Python activities.  If something told you to do that, it's wrong.  Declare any unusual dependencies in README.md for use by packaging developers..

harshkasat committed:
Feat: Integrate bs4 for html parsing to replace bs3
- added `bs4` src dir to codebase
- replacing existing parsing logic with local `BeautifulSoup` func
- refactor parsing code for improved integration
- implemented error handling with ebedded HTML parsing

harshkasat committed:
Feat: Following upate the README.md
- Project short decs
- Added license GPLv2, GPLv3
- Project overview
- Prerequisites to run infoslicer activity
- Installation Guide
- Contact and Support

harshkasat commented:
@quozl I have completed the port to py3  work on the Debian env and am currently working with Fedora.

harshkasat committed:
Feat: Port to py2 to py3
- using bs4 htmlparser
- removing prev BeautifullSoup code
- change parse from xml to html.paser
- removing trailing whitespace
- adding error handling
- use pep8 namespace
- change const name book, wiki to BOOK, WIKI

quozl commented:
Sorry, don't understand your question.

You haven't mentioned the already open pull request; https://github.com/sugarlabs/infoslicer/pull/34 can you confirm if you have used any of that work, and if not why not?

Quick view of your changes so far; you've removed a copyright notice from IBM, why?  You've removed Lauren's contact details; why?  Did you ask her?

You've vendored the entire bs4 test suite, why?

Issues on this repository include a discussion of using Python 3 html.parser rather than BeautifulSoup.  Why did you persist with BeautifulSoup?

harshkasat commented:
> You haven't mentioned the already open pull request; #34 can you confirm if you have used any of that work, and if not why not?

I used some of the code, initial after resolving each error, they might get removed.

> Quick view of your changes so far; you've removed a copyright notice from IBM, why? You've removed Lauren's contact details; why? Did you ask her?

Sorry, I forgot about that, my assumption was to redirect to the sugar matrix, but I will revert it.


> You've vendored the entire bs4 test suite, why?
> you may vendor it into the software if you give good reasons in your commit message.

the previous code was written in bs3 , so vendor bs4 entire code, If this is a wrong way, I can add bs4 as dependencies in README


> Issues on this repository include a discussion of using Python 3 html.parser rather than BeautifulSoup. Why did you persist with BeautifulSoup?

Since the previous code was coupled with Beautiful Soup, I was only using the code and making changes where Beautiful Soup was utilized. The old code was entirely written using Beautiful Soup. We can further utilities bs4 for `AI summaries and lesson plan using infoslicer activity (GSoC project)`

MostlyKIGuess commented on .gitignore:
```diff
@@ -3,4 +3,4 @@
 dist
 locale
-
+.venv
```
probably not needed, you can locally ignore this

harshkasat replied:
I will add dependencies instead of vendors in the repo.

MostlyKIGuess commented on book.py:
```diff
@@ -17,23 +17,23 @@
 
-wiki = None
-custom = None
+WIKI = None
```
I don't see how this was needed as you changed this to caps and made the similar changes everywhere

MostlyKIGuess replied:
But I guess it's fine

harshkasat replied:
better understanding of wiki and book flow,

MostlyKIGuess commented on book.py:
```diff
@@ -142,32 +150,32 @@ def __init__(self, preinstalled, root):
-            except:
-                logger.debug('cannot find index file; use empty')
+            except (IOError, json.JSONDecodeError) as e:
+                logger.debug('cannot load index file: %s', e)
```
Good logging everywhere! Thanks

MostlyKIGuess commented on book.py:
```diff
@@ -229,26 +237,26 @@ def __init__(self, filepath=None):
         else:
-            zip = zipfile.ZipFile(filepath, 'r')
-            for i in zip.namelist():
+            zip_file = zipfile.ZipFile(filepath, 'r')
```
Even here you did a similar thing as the caps earlier, keeping it zip would have been fine no?

harshkasat replied:
It's a naming convention, so I created it like this. nothing else.

MostlyKIGuess commented on bookview.py:
```diff
@@ -68,7 +67,6 @@ def __init__(self, book, name, tooltip, custom):
             title.insert(tool_item, -1)
 
         # title caption
-
```
Unnecessary

harshkasat replied:
these are small I done, to understand code.

MostlyKIGuess commented on bookview.py:
```diff
@@ -47,7 +47,6 @@ def __init__(self, book, name, tooltip, custom):
         title = Gtk.Toolbar()
 
         # title checkbox
-
```
Unnecessary

harshkasat replied:
these are small I done, to understand code.

MostlyKIGuess reviewed (commented):
Sorry I will test this again but it's very hard to find what changes you made exactly, remove the fuzz from bs4 as that's not even a code, ideally we can just add as dependencies in readme and should be fine to run with it.

harshkasat committed:
- Remove bs4 module from codebase
- Remove .venv from .gitignore
- Update Readme.md add new dependencies bs4

harshkasat committed:
Adding Copyright and License IBM Corporation 2008
Contact of Laura Cowen

harshkasat committed:
Just Rookie mistake

quozl commented:
Had another quick look at the aggregate change; still a lot of whitespace change that isn't part of porting, and Lauren is still removed.  What's up?

harshkasat commented:
I removed trailing whitespace to organize the code. Before, it was just chaos. I also added Lauren to IBM copyright and License. Should I include her under Contact and Support?

quozl commented:
I didn't find it chaos before.  It was just whitespace.  The problem is that your pull request is filled with a combination of both whitespace changes, and changes that don't appear to be related to porting to Python 3.

Yes, I know what you did, it is clear from the pull request tab "Files changed".

You haven't demonstrated a capacity to rewrite your commits, instead you're piling more commits on top of the existing ones.  I don't want to merge a commit that removes Lauren and and IBM copyright even though there is a later commit that adds the information back in a different form.  If the form must change, it should be a single commit ... and it is entirely unrelated to porting to Python 3, so it begs the question of why it is being done in a pull request titled thus.

harshkasat commented:
I understand that it may seem like just whitespace, but as we are porting to Python 3, we must adhere to PEP 8 standards and address any pylint errors. There are also several other issues, such as class names not following conventions and missing docstrings in functions. I thought we could update the docstrings according to the conventions. I can revert my changes if you don't like this approach. 

Regarding the commit, I plan to squash it once everything is functioning correctly. and I can change the pr request * Port Python3 and File Changes*

quozl commented:
I'm not saying whitespace changes are not needed, I'm saying they are getting in the way of review of your change, by hiding what you've changed inside a lot of noise.

It is almost as if you don't know the extent of your change and how those changes connect to the commit messages.  Or you haven't committed frequently enough.

One way to review while ignoring whitespace is for me to `git fetch origin pull/37/head`, `git checkout 6fdaffb`, then `git diff -w master`, which I piped to git-delta.

I've had another attempt at reviewing using this method, and again;
* [ ] the commit messages don't correspond to the changes,
* [ ] many comments are removed without explanation,
* [ ] line wrapping is changed,
* [ ] the license text is changed in README.md, (only the copyright owner should do that),
* [ ] the contact is removed in README.md, (still, despite it being mentioned several times),
* [ ] class names are changed,
* [ ] method names are changed,
* [ ] variable names are changed.

https://github.com/sugarlabs/sugar-docs/blob/master/src/contributing.md#guide-for-reviewers is our guide for reviewers, so you can see some of the things we are to think about.
